rust/src/kernel/schema.rs

```diff
@@ -0,0 +1,788 @@
+    }
+
+    #[inline]
+    /// Returns the value type of the map
+    pub const fn value_type(&self) -> &DataType {
+        &self.value_type
+    }
+
+    #[inline]
+    /// Returns whether the map can contain null values
+    pub const fn value_contains_null(&self) -> bool {
+        self.value_contains_null
+    }
+}
+
+fn default_true() -> bool {
+    true
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
+#[serde(rename_all = "camelCase")]
+/// Primitive types supported by Delta
+pub enum PrimitiveType {
+    /// UTF-8 encoded string of characters
+    String,
+    /// i64: 8-byte signed integer. Range: -9223372036854775808 to 9223372036854775807
+    Long,
+    /// i32: 4-byte signed integer. Range: -2147483648 to 2147483647
+    Integer,
+    /// i16: 2-byte signed integer numbers. Range: -32768 to 32767
+    Short,
+    /// i8: 1-byte signed integer number. Range: -128 to 127
+    Byte,
+    /// f32: 4-byte single-precision floating-point numbers
+    Float,
+    /// f64: 8-byte double-precision floating-point numbers
+    Double,
+    /// bool: boolean values
+    Boolean,
+    /// Binary: uninterpreted binary data
+    Binary,
+    /// Date: Calendar date (year, month, day)
+    Date,
+    /// Microsecond precision timestamp, adjusted to UTC.
+    Timestamp,
+    // TODO: timestamp without timezone
+    #[serde(
+        serialize_with = "serialize_decimal",
+        deserialize_with = "deserialize_decimal",
+        untagged
+    )]
+    /// Decimal: arbitrary precision decimal numbers
+    Decimal(i32, i32),
+}
+
+fn serialize_decimal<S: serde::Serializer>(
+    precision: &i32,
+    scale: &i32,
+    serializer: S,
+) -> Result<S::Ok, S::Error> {
+    serializer.serialize_str(&format!("decimal({},{})", precision, scale))
+}
+
+fn deserialize_decimal<'de, D>(deserializer: D) -> Result<(i32, i32), D::Error>
+where
+    D: serde::Deserializer<'de>,
+{
+    let str_value = String::deserialize(deserializer)?;
+    if !str_value.starts_with("decimal(") || !str_value.ends_with(')') {
+        return Err(serde::de::Error::custom(format!(
+            "Invalid decimal: {}",
+            str_value
+        )));
+    }
+
+    let mut parts = str_value[8..str_value.len() - 1].split(',');
+    let precision = parts
+        .next()
+        .and_then(|part| part.trim().parse::<i32>().ok())
+        .ok_or_else(|| {
+            serde::de::Error::custom(format!("Invalid precision in decimal: {}", str_value))
+        })?;
+    let scale = parts
+        .next()
+        .and_then(|part| part.trim().parse::<i32>().ok())
+        .ok_or_else(|| {
+            serde::de::Error::custom(format!("Invalid scale in decimal: {}", str_value))
+        })?;
+
+    Ok((precision, scale))
+}
+
+impl Display for PrimitiveType {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            PrimitiveType::String => write!(f, "string"),
+            PrimitiveType::Long => write!(f, "bigint"),
```

Comment:
Oh I think this is a bug: [Delta Protocol says "long"](https://github.com/delta-io/delta/blob/master/PROTOCOL.md#schema-serialization-format)